tests/cbtest/gtest/test_uzfs.cc

```diff
@@ -1173,3 +1174,472 @@ TEST(VolumeNameCompare, VolumeNameCompareTest) {
+	}
+
+	/* Single Replica */
+	if (mgmt_test_case == 24) {
+		hdr.opcode = ZVOL_OPCODE_START_REBUILD;
+		hdr.len = sizeof (mgmt_ack_t);
+		bzero(&mgmt_ack, sizeof (mgmt_ack));
+		strcpy(mgmt_ack.dw_volname, zinfo->name);
+		p = (char *)&mgmt_ack;
+	}
+
+send_hdr:
+	rc = write(mgmt_fd, (void *)&hdr, sizeof (zvol_io_hdr_t));
+	if (rc == -1) {
+		LOG_ERRNO("Write error during hdr write");
+		goto exit;
+	}
+
+	/*
+	 * Disconnect after send header
+	 */
+	if (mgmt_test_case == 3)
+		goto exit;
+
+	rc = write(mgmt_fd, p, hdr.len);
+	if (rc == -1) {
+		LOG_ERRNO("Write error during volname write");
+		goto exit;
+	}
+
+	rc = read(mgmt_fd, (void *)&hdr, sizeof (zvol_io_hdr_t));
+	if (rc == -1) {
+		LOG_ERRNO("Read error during hdr read");
+		goto exit;
+	}
+
+	status = hdr.status;
+	/* Reset zv in to zinfo */
+	if (mgmt_test_case == 22)
+		zinfo->zv = zv;
+
+	if (mgmt_test_case == 23)
+		uzfs_zvol_set_rebuild_status(zinfo->zv,
+		    ZVOL_REBUILDING_INIT);
+
+	if (hdr.status == ZVOL_OP_STATUS_FAILED) {
+		LOG_ERRNO("Status is not ok");
+		goto exit;
+	}
+
+	ack = (char *) malloc(hdr.len);
+	rc = read(mgmt_fd, (void *)ack, hdr.len);
+	if (rc == -1) {
+		LOG_ERRNO("Read error during ack read");
+		goto exit;
+	}
+	free(ack);
+
+	bzero(&hdr, sizeof (hdr));
+	hdr.version = REPLICA_VERSION;
+	hdr.len = strlen(zinfo->name) + 1;
+
+	if (!executed && mgmt_test_case == 8) {
+		hdr.opcode = ZVOL_OPCODE_PREPARE_FOR_REBUILD;
+		executed = true;
+		goto send_hdr;
+	}
+
+	if (!executed && mgmt_test_case == 9) {
+		hdr.opcode = ZVOL_OPCODE_REPLICA_STATUS;
+		executed = true;
+		goto send_hdr;
+	}
+		
+	if (!executed && mgmt_test_case == 10) {
+		hdr.opcode = ZVOL_OPCODE_STATS;
+		executed = true;
+		goto send_hdr;
+	}
+exit:
+	if (tgt_fd != -1)
+		close(tgt_fd);
+
+	if (mgmt_fd != -1)
+		close(mgmt_fd);
+
+	mgmt_test_case = 0;
+	zk_thread_exit();
+}
+
+void mgmt_thread_test_case(int test_case)
+{
+	kthread_t	*mock_tgt_thrd;
+	uzfs_mgmt_conn_t *conn = (uzfs_mgmt_conn_t *)zinfo->mgmt_conn;
+
+	mgmt_test_case = test_case;
+	uzfs_zinfo_take_refcnt(zinfo);
+
+	mock_tgt_thrd = zk_thread_create(NULL, 0, mock_tgt_thread,
+	    NULL, 0, NULL, TS_RUN, 0, 0);
+
+	/* wait for mock_tgt_thread to exit */
+	while (1) {
+		if (mgmt_test_case != 0)
+			sleep(1);
+		else
+			break;
+	}
+}
+
+kthread_t	*mgmt_thread;
+TEST(MgmtThreadTest, MgmtThreadCreation) {
+	EXPECT_EQ(2, uzfs_mgmt_conn_list_count(&uzfs_mgmt_conns));
+	EXPECT_EQ(2, zinfo->refcnt);
+
+	mgmt_thread = zk_thread_create(NULL, 0, uzfs_zvol_mgmt_thread,
+	    NULL, 0, NULL, TS_RUN, 0, 0);
+	sleep(1);
+	uzfs_mgmt_conn_t *conn = (uzfs_mgmt_conn_t *)zinfo->mgmt_conn;
+	conn->conn_bufsiz = 0;
+	zinfo_destroy_cb(zinfo2);
+	EXPECT_EQ(0, !mgmt_thread);
+}
+
+/* Disconnect mgmt connection just after connect */
+TEST(MgmtThreadTest, ConnBreakFromTGTAfterConnect) {
+	uzfs_mgmt_conn_t *conn = (uzfs_mgmt_conn_t *)zinfo->mgmt_conn;
+	mgmt_thread_test_case(1);
+	EXPECT_EQ(conn->conn_state, CS_READ_VERSION);
```

Comment:
if target is exited after connection, if I'm not wrong, connection should be broken/closed at replica. Right?

Reply:
Well, it start from -1, once fd is connected to tgt then it is changed to CS_CONNECT. When we add it to epoll event, it also send an event and this event change this state to CS_READ_VERSION i.e. connection waiting for reading hdr version. Now if you disconnect tgt, we  end up here, we do not set it to CS_CLOSE in close_conn.